### ISSUE TITLE
cdist and extract_iter functions give slightly different results

### ISSUE THREAD

**Issue by NikhilKothari** [maxbachmann/rapidfuzz#132]
@maxbachmann Thanks for the awesomeness that RapidFuzz is!
```
q1 = ["LIFE"]
q2 = ["MYSTERY"]

process.cdist(q1, q2, scorer=fuzz.partial_ratio)
>>  array([[28]], dtype=uint8)

list(process.extract_iter(q1[0], q2, scorer=fuzz.partial_ratio)
>> [('MYSTERY', 28.57142857142857, 0)]
```
Is there a way to get the `cdist` result in floating-point, exactly similar to `extract_iter`? (_extract_iter works slightly better in my use case, but cdist is way faster!)_

**Comment by maxbachmann** [comment]
1) This result is incorrect. It should be 29 after rounding. 
2) I currently use the type uint8_t, since it uses significantly less memory than double (which becomes relevant for large matrixes).

Currently it is not possible to change this type. However I could add an additional argument, which allows the selection in the following way:
```python
>>> process.cdist(q1, q2, scorer=fuzz.partial_ratio, dtype=np.double)
array([[28.57142857142857]], dtype=double)
```

> cdist is way faster!

It is definetly recommended to use `cdist`. As you noted it is already very fast today and some of the future improvements like simd and multithreading will especially improve the performance of `cdist`.

**Comment by NikhilKothari** [comment]
```
>>> process.cdist(q1, q2, scorer=fuzz.partial_ratio, dtype=np.double)
array([[28.57142857142857]], dtype=double)
```
This addition would be really helpful! Looking forward.

**Comment by maxbachmann** [comment]
Added support for this in the latest version on main (will be part of the next release):
```python
>>> import numpy as np
>>> from rapidfuzz import process, fuzz
>>> q1 = ["LIFE"]
>>> q2 = ["MYSTERY"]
>>> process.cdist(q1, q2, scorer=fuzz.partial_ratio)
array([[29]], dtype=uint8)
>>> process.cdist(q1, q2, scorer=fuzz.partial_ratio, dtype=np.float32)
array([[28.571428]], dtype=float32)
>>> process.cdist(q1, q2, scorer=fuzz.partial_ratio, dtype=np.float64)
array([[28.57142857]])
```
for now this supports different dtypes depending on the scorer.
For similarities like fuzz.ratio:
- np.uint8 (default)
- np.float32
- np.float64/np.double

For distances like string_metric.levenshtein:
- np.int8
- np.int16
- np.int32 (default)
- np.int64

**Comment by NikhilKothari** [comment]
Super! Thanks, @maxbachmann!

**Comment by maxbachmann** [comment]
@NikhilKothari this is available now. In addition cdist supports multithreading now to further improve the performance ;)
```python
process.cdist(q1, q2, scorer=fuzz.partial_ratio, dtype=np.float32, workers=-1)
```

**Comment by NikhilKothari** [comment]
I've been checking for the new release every day! :D
Thanks man, I'm trying it out right now; will let you know soon.

**Comment by NikhilKothari** [comment]
Hey @maxbachmann!

**TL;DR**
This could just be an update for you and might need no action, but still the information could come in handy.
On a K80 GPU with 4 Core CPU and 61GB RAM, `cdist` in version 1.7.0 (workers=-1, dtype=np.float32) performs better at 3k RPM load, but at 6k RPM `cdist` from version 1.6.2 seems to perform slightly better.

_Note_ - I'm using Rapidfuzz in conjunction with Pytorch so the numbers below wouldn't be just for `cdist` but for the end to end process. For comparison purposes, I've kept everything else the same, just the `cdist` functions from 1.6.2 and 1.7.0 versions different. The numbers below are in milliseconds.

**Version 1.6.2**
*3k RPM Load*
![1 6 2_3k](https://user-images.githubusercontent.com/19628452/135235366-ce5778ef-0a8b-4405-a3a4-846bd629ac1d.png)

*6k RPM Load*
![1 6 2_6k](https://user-images.githubusercontent.com/19628452/135235481-fdf163dc-22e5-4108-b3a6-13e03fd82744.png)

**Version 1.7.0**
*3k RPM Load*
![1 7 0_3k](https://user-images.githubusercontent.com/19628452/135235673-dad6b7a9-3027-4657-adb6-7c980725ca18.png)

*6k RPM Load*
![1 7 0_6k](https://user-images.githubusercontent.com/19628452/135235749-fb6ba257-28a4-4aab-aeb7-c08eb8d581ac.png)


This issue is not very significant but I still thought that you should know. `cdist` is anyway a huge improvement over `extract_iter` and is golden for all use cases. One more thing to note here about my use case and probably the cause of this issue is that the collections I'm using are not big, just a list of 150 strings of length 15 words each. The improvement might have been more evident if the collections were big. Also, these load testing numbers could have been a one-off event anomaly; I'm running more just to be sure.

**Comment by maxbachmann** [comment]
> One more thing to note here about my use case and probably the cause of this issue is that the collections I'm using are not big, just a list of 150 strings of length 15 words each

When working with a small amount of elements, you probably should not use the multithreaded version. For small counts, the overhead caused by creating the threads + the work queues might be bigger than the performance improvement it achieves. 

When using a single worker `worker=1` I would expect the two versions to behave relatively similar.

**Comment by NikhilKothari** [comment]
Makes sense, I'll try it out.
